### PR TITLE
spack find external support for OpenCV and OpenBLAS

### PR DESCRIPTION
**Pull request by bvanessen**
Added support for finding the OpenCV package via the find external
command.  Included support for identifying variants based on available
shared libraries.

Added support to finding the OpenBLAS package via the find external
command.

Enabled pacakges to show that they can be discovered via the find
external command in the info message.

### CODE REVIEW

**Comment by bvanessen** [comment]
@alalazo If you get a chance to look over this PR tonight, that would be great.

**Review by alalazo — APPROVED** [review]
LGTM. Code for finding executables and libraries is very similar, if not equal, in most packages. We should probably do something to gather it together in a core function that can be reused, but that's for a later PR.

File: var/spack/repos/builtin/packages/opencv/package.py [hunk 1]
```diff
@@ -233,10 +234,16 @@ class Opencv(CMakePackage, CudaPackage):
...
 
+    # Define the list of libraries objects that may be used
+    # to find an external installation and its variants
+    libraries = []
+
     # module variants
     for mod in modules:
         # At least one of these modules must be enabled to build OpenCV
         variant(mod, default=False, description="Include opencv_{0} module".format(mod))
+        lib = f'libopencv_{mod}'
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/opencv/package.py]: Can't use f-strings, we need to support Python 2.7

**Review by alalazo — CHANGES_REQUESTED** [review]
I'll merge as soon as tests pass :slightly_smiling_face:

**Comment by adamjstewart** [comment]
Can this be split into 2 separate PRs? These 2 packages are independent of each other.

File: var/spack/repos/builtin/packages/opencv/package.py (hunk 1 above)
Comment (bvanessen) [reply var/spack/repos/builtin/packages/opencv/package.py]: Okay, I just pushed a fix.

**Review by alalazo — DISMISSED** [review]

**Comment by alalazo** [comment]
> Can this be split into 2 separate PRs? These 2 packages are independent of each other.

I think that's a good advice in general, but let's not add a burden on the submitter for this time :slightly_smiling_face:

**Review by adamjstewart — CHANGES_REQUESTED** [review]
These won't work on macOS which uses `.dylib` instead of `.so`. Here are some example file names:
```console
$ spack find -pv opencv
opencv@4.5.4~1394~ade~alphamat~android_mediandk~android_native_camera~annotation~aruco~avfoundation~barcode~bgsegm~bioinspired~calib3d~cap_ios~carotene~ccalib~clp~cpufeatures~createsamples~cublas~cuda~cudaarithm~cudabgsegm~cudacodec~cudafeatures2d~cudafilters~cudaimgproc~cudalegacy~cudaobjdetect~cudaoptflow~cudastereo~cudawarping~cudev~cudnn~cufft~cvv~datasets~directx~dnn~dnn_objdetect~dnn_superres~dpm~dshow~eigen~face~fast-math~features2d~ffmpeg~flann~freetype~fuzzy~gapi~gdal~gtk~hdf~hfs~highgui~hpx~img_hash~imgcodec_hdr~imgcodec_pfm~imgcodec_pxm~imgcodec_sunraster+imgcodecs+imgproc~intensity_transform~interactive-calibration~ipo~ipp~itt~jasper~java~java_bindings_generator+jpeg~lapack~line_descriptor~matlab~mcc~ml~model-diagnostics~msmf~msmf_dxva~nonfree~objc~objc_bindings_generator~objdetect~onnx~opencl~opencl_d3d11_nv~openexr~opengl~openjpeg~openmp~optflow~phase_unwrapping~photo~plaidml~plot+png~powerpc~protobuf~pthreads_pf~python2+python3+python_bindings_generator~python_tests~qt~quality~quirc~rapid~reg~rgbd~saliency~sfm~shape+shared~stereo~stitching~structured_light~superres~surface_matching~tbb~tengine~tesseract~text+tiff~tracking~traincascade~ts~v4l~version~video~videoio~videostab~visualisation~viz~vtk~vulcan~webp~wechat_qrcode~win32ui~world~xfeatures2d~ximgproc~xobjdetect~xphoto build_type=RelWithDebInfo patches=1d30f12,2aea289,3cab451  /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/opencv-4.5.4-a6ggmspsgsc6cvelpnfqufec5c6s76xt
$ ls /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/opencv-4.5.4-a6ggmspsgsc6cvelpnfqufec5c6s76xt/lib/
cmake				libopencv_core.dylib		libopencv_imgcodecs.dylib	libopencv_imgproc.dylib
libopencv_core.4.5.4.dylib	libopencv_imgcodecs.4.5.4.dylib	libopencv_imgproc.4.5.4.dylib	python3.9
libopencv_core.4.5.dylib	libopencv_imgcodecs.4.5.dylib	libopencv_imgproc.4.5.dylib
```

**Comment by bvanessen** [comment]
> These won't work on macOS which uses `.dylib` instead of `.so`. Here are some example file names:
> 
> ```
> $ spack find -pv opencv
> opencv@4.5.4~1394~ade~alphamat~android_mediandk~android_native_camera~annotation~aruco~avfoundation~barcode~bgsegm~bioinspired~calib3d~cap_ios~carotene~ccalib~clp~cpufeatures~createsamples~cublas~cuda~cudaarithm~cudabgsegm~cudacodec~cudafeatures2d~cudafilters~cudaimgproc~cudalegacy~cudaobjdetect~cudaoptflow~cudastereo~cudawarping~cudev~cudnn~cufft~cvv~datasets~directx~dnn~dnn_objdetect~dnn_superres~dpm~dshow~eigen~face~fast-math~features2d~ffmpeg~flann~freetype~fuzzy~gapi~gdal~gtk~hdf~hfs~highgui~hpx~img_hash~imgcodec_hdr~imgcodec_pfm~imgcodec_pxm~imgcodec_sunraster+imgcodecs+imgproc~intensity_transform~interactive-calibration~ipo~ipp~itt~jasper~java~java_bindings_generator+jpeg~lapack~line_descriptor~matlab~mcc~ml~model-diagnostics~msmf~msmf_dxva~nonfree~objc~objc_bindings_generator~objdetect~onnx~opencl~opencl_d3d11_nv~openexr~opengl~openjpeg~openmp~optflow~phase_unwrapping~photo~plaidml~plot+png~powerpc~protobuf~pthreads_pf~python2+python3+python_bindings_generator~python_tests~qt~quality~quirc~rapid~reg~rgbd~saliency~sfm~shape+shared~stereo~stitching~structured_light~superres~surface_matching~tbb~tengine~tesseract~text+tiff~tracking~traincascade~ts~v4l~version~video~videoio~videostab~visualisation~viz~vtk~vulcan~webp~wechat_qrcode~win32ui~world~xfeatures2d~ximgproc~xobjdetect~xphoto build_type=RelWithDebInfo patches=1d30f12,2aea289,3cab451  /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/opencv-4.5.4-a6ggmspsgsc6cvelpnfqufec5c6s76xt
> $ ls /Users/Adam/spack/opt/spack/darwin-catalina-ivybridge/apple-clang-12.0.0/opencv-4.5.4-a6ggmspsgsc6cvelpnfqufec5c6s76xt/lib/
> cmake				libopencv_core.dylib		libopencv_imgcodecs.dylib	libopencv_imgproc.dylib
> libopencv_core.4.5.4.dylib	libopencv_imgcodecs.4.5.4.dylib	libopencv_imgproc.4.5.4.dylib	python3.9
> libopencv_core.4.5.dylib	libopencv_imgcodecs.4.5.dylib	libopencv_imgproc.4.5.dylib
> ```

@adamjstewart Okay I have added support for finding OS X based dynamic libraries.

**Comment by bvanessen** [comment]
@tgamblin @becker33 @scheibelp Since this now touches a function in the main part of Spack, I didn't know if you would want to look at it.

**Comment by bvanessen** [comment]
@adamjstewart or @alalazo Is there anything more on this PR that you want to see?

File: lib/spack/llnl/util/filesystem.py [hunk 2]
```diff
@@ -109,12 +110,17 @@ def path_contains_subdirectory(path, root):
     return norm_path.startswith(norm_root)
 
 
+def library_extensions():
```
Comment (adamjstewart) [inline lib/spack/llnl/util/filesystem.py]: This should prob be a variable instead of a function

File: lib/spack/spack/detection/path.py [hunk 3]
```diff
@@ -87,7 +87,9 @@ def libraries_in_ld_library_path(path_hints=None):
         path_hints (list): list of paths to be searched. If None the list will be
             constructed based on the LD_LIBRARY_PATH environment variable.
     """
-    path_hints = path_hints or spack.util.environment.get_path('LD_LIBRARY_PATH')
+    path_hints = path_hints or \
+        spack.util.environment.get_path('LD_LIBRARY_PATH') or \
+        spack.util.environment.get_path('DYLD_LIBRARY_PATH')
```
Comment (adamjstewart) [inline lib/spack/spack/detection/path.py]: Why do we only search `LD_LIBRARY_PATH` and not `LIBRARY_PATH`? Also, you should add `DYLD_FALLBACK_LIBRARY_PATH`, this is what Spack uses.

**Review by adamjstewart — COMMENTED** [review]
Still need to actually test the code but here is a first pass review

File: lib/spack/llnl/util/filesystem.py (hunk 2 above)
Comment (bvanessen) [reply lib/spack/llnl/util/filesystem.py]: Done

File: lib/spack/spack/detection/path.py (hunk 3 above)
Comment (bvanessen) [reply lib/spack/spack/detection/path.py]: Done.

File: lib/spack/llnl/util/filesystem.py [hunk 4]
```diff
@@ -109,12 +110,16 @@ def path_contains_subdirectory(path, root):
     return norm_path.startswith(norm_root)
 
 
+"""This generates the library filenames that may appear on any OS.
+"""
```
Comment (adamjstewart) [inline lib/spack/llnl/util/filesystem.py]:
```suggestion
#: This generates the library filenames that may appear on any OS.
```

**Comment by bvanessen** [comment]
@alalazo @adamjstewart I just rebased the PR to make it easier to merge.

File: var/spack/repos/builtin/packages/openblas/package.py [hunk 5]
```diff
@@ -6,6 +6,8 @@
 import os
 import re
 
+from llnl.util import filesystem
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/openblas/package.py]: I would just use `from llnl.util.filesystem import library_extensions`

**Comment by adamjstewart** [comment]
Tried `spack load opencv` and `spack external find opencv` but it didn't find the installation even though it's in my `DYLD_FALLBACK_LIBRARY_PATH`. Do you have access to a macOS machine to test this? If not I can take a closer look and try to figure out why it's not working.

**Comment by bvanessen** [comment]
> Tried `spack load opencv` and `spack external find opencv` but it didn't find the installation even though it's in my `DYLD_FALLBACK_LIBRARY_PATH`. Do you have access to a macOS machine to test this? If not I can take a closer look and try to figure out why it's not working.

@adamjstewart This is what worked for me on OS X `LD_LIBRARY_PATH=/usr/local/Cellar/openblas/0.3.20/ spack -d external find openblas`

File: var/spack/repos/builtin/packages/openblas/package.py [hunk 6]
```diff
@@ -145,6 +149,17 @@ class Openblas(MakefilePackage):
 
     depends_on('perl', type='build')
 
+    @classmethod
+    def determine_version(cls, lib):
+        ver = None
+        lib_extensions = library_extensions
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/openblas/package.py]: I don't think this variable is needed, we can just use `library_extensions` directly.

**Comment by adamjstewart** [comment]
I also tried setting `LD_LIBRARY_PATH` to the same value as `DYLD_FALLBACK_LIBRARY_PATH` but `spack external find opencv` still doesn't pick it up. Can you try installing OpenCV with Spack and see if `spack load opencv && spack external find opencv` works for you? Maybe Homebrew uses a different installation layout?

File: var/spack/repos/builtin/packages/openblas/package.py (hunk 5 above)
Comment (bvanessen) [reply var/spack/repos/builtin/packages/openblas/package.py]: Done

File: var/spack/repos/builtin/packages/openblas/package.py (hunk 6 above)
Comment (bvanessen) [reply var/spack/repos/builtin/packages/openblas/package.py]: Done

**Comment by bvanessen** [comment]
> I also tried setting `LD_LIBRARY_PATH` to the same value as `DYLD_FALLBACK_LIBRARY_PATH` but `spack external find opencv` still doesn't pick it up. Can you try installing OpenCV with Spack and see if `spack load opencv && spack external find opencv` works for you? Maybe Homebrew uses a different installation layout?

Can you try again, I think that I have fixed this.

**Comment by adamjstewart** [comment]
Thanks, that seems to have fixed it!

Another possible bug I've found. When there are multiple copies of OpenCV in `*LIBRARY_PATH`, Spack only adds a single one to my `packages.yaml`. This is quite different than how `spack external find` works for `PATH` when searching for binaries. Is there some logic that leaves early once one library is found?

Also, I noticed that the variant detection adds `+foo` for all libraries that are found, but doesn't add `~bar` for libraries that aren't found. I think it would be helpful to add `~bar` too.

**Comment by bvanessen** [comment]
> Thanks, that seems to have fixed it!
> 
> Another possible bug I've found. When there are multiple copies of OpenCV in `*LIBRARY_PATH`, Spack only adds a single one to my `packages.yaml`. This is quite different than how `spack external find` works for `PATH` when searching for binaries. Is there some logic that leaves early once one library is found?
> 
> Also, I noticed that the variant detection adds `+foo` for all libraries that are found, but doesn't add `~bar` for libraries that aren't found. I think it would be helpful to add `~bar` too.

@adamjstewart I don’t know why it only finds one.  I didn’t do anything specifically to stop after the first.  I can check.  I didn’t know if we would want to constrain it with negatives on the features if they weren’t found, but I guess that I could.

**Comment by bvanessen** [comment]
@alalazo Do you have any idea of why only one library would be returned?

File: var/spack/repos/builtin/packages/opencv/package.py [hunk 7]
```diff
@@ -873,6 +883,43 @@ class Opencv(CMakePackage, CudaPackage):
...
+                    pattern = re.compile(r'lib(\S*?)_(\S*)\.(\d+\.\d+\.\d+)\.%s' %
+                                         ext)
+                else:
+                    pattern = re.compile(r'lib(\S*?)_(\S*)\.%s\.(\d+\.\d+\.\d+)' %
+                                         ext)
+                match = pattern.search(lib)
+                if match and not match.group(2) == 'core':
+                    variants.append('+' + match.group(2))
+        results.append(' '.join(variants))
+        return results
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/opencv/package.py]: Shouldn't this be:
```suggestion
        return ' '.join(variants)
```
?

File: var/spack/repos/builtin/packages/opencv/package.py [hunk 8]
```diff
@@ -873,6 +883,43 @@ class Opencv(CMakePackage, CudaPackage):
...
+                pattern = re.compile(r'lib(\S*?)_(\S*)\.%s\.(\d+\.\d+\.\d+)' %
+                                     ext)
+            match = pattern.search(lib)
+            if match:
+                ver = match.group(3)
+        return ver
+
+    @classmethod
+    def determine_variants(cls, libs, version_str):
+        results = []
```
Comment (alalazo) [inline var/spack/repos/builtin/packages/opencv/package.py]:
```suggestion
```

**Review by alalazo — CHANGES_REQUESTED** [review]

**Review by alalazo — DISMISSED** [review]

**Comment by bvanessen** [comment]
> > Thanks, that seems to have fixed it!
> > Another possible bug I've found. When there are multiple copies of OpenCV in `*LIBRARY_PATH`, Spack only adds a single one to my `packages.yaml`. This is quite different than how `spack external find` works for `PATH` when searching for binaries. Is there some logic that leaves early once one library is found?
> > Also, I noticed that the variant detection adds `+foo` for all libraries that are found, but doesn't add `~bar` for libraries that aren't found. I think it would be helpful to add `~bar` too.
> 
> @adamjstewart I don’t know why it only finds one. I didn’t do anything specifically to stop after the first. I can check. I didn’t know if we would want to constrain it with negatives on the features if they weren’t found, but I guess that I could.

@adamjstewart we have a fix that has been applied.  Thanks to @alalazo for identifying and producing the patch.

**Comment by bvanessen** [comment]
@adamjstewart I added the support to mark the variants not used.

**Review by adamjstewart — APPROVED** [review]
Confirmed that this works on macOS. It picks up my OpenCV installation and correctly determines the version and both positive and negative variants.

However, it's still not picking up multiple installations when they are all in my `*LIBRARY_PATH`. But that was already broken, so I'm fine with merging this as is and fixing that in a separate PR if you want.

**Review by alalazo — APPROVED** [review]

**Pull request merged**